lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs

```diff
@@ -413,6 +419,187 @@ spec = do
+
+        it "epochStartTime (epochCeiling t) >= t" $
+            withMaxSuccess 1000 $ property $
+                \(SlotParametersAndTimePoint sps time) -> do
+                    let timeMaximum = epochStartTime sps maxBound
+                    let withinBounds = time <= timeMaximum
+                    checkCoverage $
+                        cover 10      withinBounds  "within bounds" $
+                        cover 10 (not withinBounds) "out of bounds" $
+                        case epochCeiling sps time of
+                            Nothing -> not withinBounds
+                            Just en -> time <= epochStartTime sps en
+
+        it "epochStartTime (epochPred (epochCeiling t)) < t" $
+            withMaxSuccess 1000 $ property $
+                \(SlotParametersAndTimePoint sps time) -> do
+                    let timeMaximum = epochStartTime sps maxBound
+                    let withinBounds = time <= timeMaximum
+                    checkCoverage $
+                        cover 10      withinBounds  "within bounds" $
+                        cover 10 (not withinBounds) "out of bounds" $
+                        case epochCeiling sps time of
+                            Nothing -> not withinBounds
+                            Just e1 -> case epochPred e1 of
+                                Nothing -> e1 == minBound
+                                Just e2 -> time > epochStartTime sps e2
+
+    describe "Epoch arithmetic: epochFloor: core properties" $ do
+
+        it "epochStartTime (epochFloor t) <= t" $
+            withMaxSuccess 1000 $ property $
+                \(SlotParametersAndTimePoint sps time) -> do
+                    let timeMinimum = epochStartTime sps minBound
+                    let withinBounds = time >= timeMinimum
+                    checkCoverage $
+                        cover 10      withinBounds  "within bounds" $
+                        cover 10 (not withinBounds) "out of bounds" $
+                        case epochFloor sps time of
+                            Nothing -> not withinBounds
+                            Just en -> time >= epochStartTime sps en
+
+        it "epochStartTime (epochSucc (epochFloor t)) > t" $
+            withMaxSuccess 1000 $ property $
+                \(SlotParametersAndTimePoint sps time) -> do
+                    let timeMinimum = epochStartTime sps minBound
+                    let withinBounds = time >= timeMinimum
+                    checkCoverage $
+                        cover 10      withinBounds  "within bounds" $
+                        cover 10 (not withinBounds) "out of bounds" $
+                        case epochFloor sps time of
+                            Nothing -> not withinBounds
+                            Just e1 -> case epochSucc e1 of
+                                Nothing -> e1 == maxBound
+                                Just e2 -> time < epochStartTime sps e2
```

Comment:
**[question]**

It seems the structure of the above 4 tests are almost the same? No chance to reuse one parametrize property function?

Reply:
You're right, they are very similar. :+1: 

I'll simplify these in a future PR.